lib/bootsnap/cli.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require 'bootsnap'
+require 'optparse'
+require 'fileutils'
+
+module Bootsnap
+  class CLI
+    unless Regexp.method_defined?(:match?)
+      module RegexpMatchBackport
+        refine Regepx do
+          def match?(string)
+            !!match(string)
+          end
+        end
+      end
+      using RegexpMatchBackport
+    end
+
+    attr_reader :cache_dir, :argv
+
+    attr_accessor :compile_gemfile, :exclude
+
+    def initialize(argv)
+      @argv = argv
+      self.cache_dir = 'tmp/cache'
+      self.compile_gemfile = false
+      self.exclude = nil
+    end
+
+    def precompile_command(*sources)
+      require 'bootsnap/compile_cache/iseq'
+
+      sources.each do |source|
+        if !File.directory?(source) && File.exist?(source)
+          STDERR.puts "#{source} is not a directory"
+          return 1
+        end
+      end
+
+      if compile_gemfile
+        sources += $LOAD_PATH
+      end
+
+      CompileCache::ISeq.compile_option_updated
+      iseq_cache_dir = File.join(cache_dir, 'bootsnap-compile-cache')
+
+      source.map { |d| File.expand_path(d) }.each do |path|
+        Dir[File.join(path, '**/*.rb')].each do |ruby_file|
+          if !exclude || !exclude.match?(ruby_file)
```

Comment:
Or even just use `=~` which does allocate a MatchData if that's what we're going for.